### PR TITLE
G-API: GOpaque implementation

### PR DESCRIPTION
**Pull request by smirnov-alexey**
This pullrequest adds a new type in G-API - GOpaque which is designed to store a single custom user type.

Currently it's almost copypasted GArray except other internal types are not related to vector.

<cut/>

```
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

force_builders=Custom,Custom Win,ARMv7
buildworker:Custom=linux-1
#build_image:Custom=ubuntu-clang:18.04
build_image:Custom=centos:7
build_contrib:Custom=OFF
```

### CODE REVIEW

**Comment by smirnov-alexey** [comment]
@TolyaTalamanov, @AsyaPronina, @anton-potapov. Any feedback is appreciated

**Comment by smirnov-alexey** [comment]
Adding GOpaque to variant broke the ABI

**Comment by alalek** [comment]
> Removed Symbols  121 

Unfortunately there is no reliable way to maintain ABI checker for G-API anymore.
We need to define long term solution.

**Comment by smirnov-alexey** [comment]
@rgarnov, @AsyaPronina, @TolyaTalamanov, please, review. As discussed with @dmatveev, 'copypaste' of GArray is OK for now, so it can be reviewed and merged.

File: modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp [hunk 1]
```diff
@@ -145,12 +150,31 @@ template<typename U> struct get_in<cv::GArray<U> >
...
+    static const U& get(GCPUContext &ctx, int idx) { return ctx.inArg<OpaqueRef>(idx).rref<U>(); }
+};
 
 //FIXME(dm): GArray<Mat>/GArray<GMat> conversion should be done more gracefully in the system
 template<> struct get_in<cv::GArray<cv::GMat> >: public get_in<cv::GArray<cv::Mat> >
 {
 };
 
+//FIXME(dm): GArray<Scalar>/GArray<GScalar> conversion should be done more gracefully in the system
+template<> struct get_in<cv::GArray<cv::GScalar> >: public get_in<cv::GArray<cv::Scalar> >
```
Comment (smirnov-alexey) [inline modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp]: I saw a comment somewhere in the code about `GArray<GMat>` -> `GArray<cv::Mat>` and `GArray<GScalar>` -> `GArray<cv::Scalar>`, so I've decided to add it

File: modules/gapi/include/opencv2/gapi/gcompoundkernel.hpp [hunk 2]
```diff
@@ -65,6 +65,16 @@ template<typename U> struct get_compound_in<cv::GArray<U>>
     }
 };
 
+template<typename U> struct get_compound_in<cv::GOpaque<U>>
```
Comment (smirnov-alexey) [inline modules/gapi/include/opencv2/gapi/gcompoundkernel.hpp]: GArray was here, so I think we can handle GOpaque too. But not sure if was needed

File: modules/gapi/include/opencv2/gapi/gtype_traits.hpp [hunk 3]
```diff
@@ -100,6 +112,7 @@ namespace detail
     template<>           struct GTypeOf<cv::gapi::own::Mat>    { using type = cv::GMat;      };
     template<>           struct GTypeOf<cv::gapi::own::Scalar> { using type = cv::GScalar;   };
     template<typename U> struct GTypeOf<std::vector<U> >       { using type = cv::GArray<U>; };
+    template<typename U> struct GTypeOf                        { using type = cv::GOpaque<U>; }; // fixme: howto differ from line 119???
```
Comment (smirnov-alexey) [inline modules/gapi/include/opencv2/gapi/gtype_traits.hpp]: Remove comment

File: modules/gapi/include/opencv2/gapi/gtype_traits.hpp [hunk 4]
```diff
@@ -165,6 +178,11 @@ namespace detail
     template<typename T> using wrap_gapi_helper = WrapValue<typename std::decay<T>::type>;
     template<typename T> using wrap_host_helper = WrapValue<typename std::decay<g_type_of_t<T> >::type>;
 
+    //if (std::is_same(wrap_gapi_helper, wrap_host_helper))
```
Comment (smirnov-alexey) [inline modules/gapi/include/opencv2/gapi/gtype_traits.hpp]: Remove it

File: modules/gapi/include/opencv2/gapi/ocl/goclkernel.hpp [hunk 5]
```diff
@@ -68,9 +68,14 @@ class GAPI_EXPORTS GOCLContext
     {
         return outVecRef(output).wref<T>();
     }
+    template<typename T> T& outOpR(int output) // FIXME: the same issue
```
Comment (smirnov-alexey) [inline modules/gapi/include/opencv2/gapi/ocl/goclkernel.hpp]: GArray was here, so I think we can handle GOpaque too. But not sure if was needed

File: modules/gapi/test/gapi_opaque_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+        Foo& operator=(Foo&& other) noexcept
+        {
+            std::swap(cstring, other.cstring);
+            return *this;
+        }
+    };
+
+    // compare 2 instances of Foo
+    auto foo_are_eq = [](const Foo& a, const Foo& b, std::size_t n){
```
Comment (smirnov-alexey) [inline modules/gapi/test/gapi_opaque_tests.cpp]: I think this can be done in a more elegant way

File: modules/gapi/test/gapi_opaque_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+        return strncmp(a.cstring, b.cstring, n);
+    };
```
Comment (dmatveev) [inline modules/gapi/test/gapi_opaque_tests.cpp]: Why craft this when we could use a regular std::string to check move semantics?

File: modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp [hunk 8]
```diff
@@ -94,9 +94,14 @@ class GAPI_EXPORTS GCPUContext
     {
         return outVecRef(output).wref<T>();
     }
+    template<typename T> T& outOpR(int output) // FIXME: the same issue
```
[the remark this answers is not in the document]
Comment (smirnov-alexey) [reply modules/gapi/include/opencv2/gapi/cpu/gcpukernel.hpp]: Renamed, thanks

File: modules/gapi/include/opencv2/gapi/gcall.hpp [hunk 9]
```diff
@@ -57,6 +63,8 @@ class GAPI_EXPORTS GCall final
 
     // Public version returns a typed array, this one is implementation detail
     detail::GArrayU yieldArray(int output = 0);
+    // Public version returns a typed opaque, this one is implementation detail
```
[the remark this answers is not in the document]
Comment (smirnov-alexey) [reply modules/gapi/include/opencv2/gapi/gcall.hpp]: Moved to the comment above

File: modules/gapi/include/opencv2/gapi/gopaque.hpp [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+
+    // FIXME: garray.hpp already contains commented hint classes below,
+    // need to think where it can be moved (currently opaque uses it from garray)
+
+    // This is the base struct for GOpaqueU type holder
+    //struct TypeHintBaseOp{virtual ~TypeHintBaseOp() = default;};
+
+    // This class holds type of initial GOpaque to be checked from GOpaqueU
+    //template <typename T>
+    //struct TypeHintOp final : public TypeHintBaseOp{};
```
[the remark this answers is not in the document]
Comment (smirnov-alexey) [reply modules/gapi/include/opencv2/gapi/gopaque.hpp]: Removed, thanks

File: modules/gapi/test/gapi_opaque_tests.cpp [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+
+G_TYPED_KERNEL(FillMat, <GMat(cv::GOpaque<int>, int, int, cv::Size)>, "test.opaque.fill_mat")
+{
+    static GMatDesc outMeta(const GOpaqueDesc&, int depth, int chan, cv::Size size)
+    {
+        return cv::GMatDesc{depth, chan, size};
+    }
+};
+
+G_TYPED_KERNEL(PaintPoint, <GMat(GPointOpaque, int, int, cv::Size)>, "test.opaque.paint_point")
```
[the remark this answers is not in the document]
Comment (smirnov-alexey) [reply modules/gapi/test/gapi_opaque_tests.cpp]: It works fine without `using`, so there're no problems here

File: modules/gapi/test/gapi_opaque_tests.cpp (hunk 7 above)
Comment (smirnov-alexey) [reply modules/gapi/test/gapi_opaque_tests.cpp]: True, it was over-complicated, fixed, thanks

**Comment by dmatveev** [comment]
> Unfortunately there is no reliable way to maintain ABI checker for G-API anymore.
> We need to define long term solution.

@alalek I thought ABI compatibility requirements were relaxed already since 4.0?

**Comment by alalek** [comment]
We use "ABI checker" tool in "source" compatibility mode for OpenCV 4.x. OpenCV 3.4 uses full binary/source checks.

**Comment by dmatveev** [comment]
https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/24319/steps/Compare%20ABI%20dumps/logs/report-html

This thing just can't into template changes :) I believe source compatibility is actually preserved here.

**Comment by alalek** [comment]
There is some issue with ARMv7 builder (GCC 4.8). My guess, it is related to `unique_ptr`.

File: modules/gapi/include/opencv2/gapi/gopaque.hpp [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        template<typename T> inline void check() const
+        {
+            GAPI_DbgAssert(dynamic_cast<OpaqueRefT<T>*>(m_ref.get()) != nullptr);
+        }
+
+    public:
+        OpaqueRef() = default;
+        template<typename T> explicit OpaqueRef(const T&  obj) : m_ref(new OpaqueRefT<T>(obj)) {}
+        template<typename T> explicit OpaqueRef(      T&  obj) : m_ref(new OpaqueRefT<T>(obj)) {}
+        template<typename T> explicit OpaqueRef(      T&& obj) : m_ref(new OpaqueRefT<T>(obj)) {}
```
Comment (alalek) [inline modules/gapi/include/opencv2/gapi/gopaque.hpp]: Removal of `T&&` case + adding specialization for `WrapValue<cv::detail::OpaqueRef>` helps with compilation on GCC 4.8

Comment (alalek) [reply modules/gapi/include/opencv2/gapi/gopaque.hpp]: > removal of `T&&`

You may try `std::remove_reference` instead:

```
        template<typename T> explicit OpaqueRef(const typename std::remove_reference<T>::type&  obj) : m_ref(new OpaqueRefT<T>(obj)) {}
        template<typename T> explicit OpaqueRef(      typename std::remove_reference<T>::type&  obj) : m_ref(new OpaqueRefT<T>(obj)) {}
        template<typename T> explicit OpaqueRef(      typename std::remove_reference<T>::type&& obj) : m_ref(new OpaqueRefT<T>(obj)) {}
```

---

<details>

<summary>Error message</summary>

```
build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/util/any.hpp:69:59: error: call of overloaded 'OpaqueRef(cv::detail::OpaqueRef&)' is ambiguous
          holder_impl(arg_t&& a) : v(std::forward<arg_t>(a)) {}
                                                           ^
/build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/util/any.hpp:69:59: note: candidates are:
In file included from /build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/garg.hpp:23:0,
                 from /build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/gcall.hpp:11,
                 from /build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/gkernel.hpp:20,
                 from /build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/core.hpp:19,
                 from /build/precommit_armv7/opencv/modules/gapi/src/precomp.hpp:14,
                 from /build/precommit_armv7/opencv/modules/gapi/src/backends/cpu/gcpubackend.cpp:8:
/build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/gopaque.hpp:222:39: note: cv::detail::OpaqueRef::OpaqueRef(T&&) [with T = cv::detail::OpaqueRef&]
         template<typename T> explicit OpaqueRef(      T&& obj) : m_ref(new OpaqueRefT<T>(obj)) {}
                                       ^
/build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/gopaque.hpp:221:39: note: cv::detail::OpaqueRef::OpaqueRef(T&) [with T = cv::detail::OpaqueRef]
         template<typename T> explicit OpaqueRef(      T&  obj) : m_ref(new OpaqueRefT<T>(obj)) {}
                                       ^
/build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/gopaque.hpp:220:39: note: cv::detail::OpaqueRef::OpaqueRef(const T&) [with T = cv::detail::OpaqueRef]
         template<typename T> explicit OpaqueRef(const T&  obj) : m_ref(new OpaqueRefT<T>(obj)) {}
                                       ^
/build/precommit_armv7/opencv/modules/gapi/include/opencv2/gapi/gopaque.hpp:209:11: note: cv::detail::OpaqueRef::OpaqueRef(const cv::detail::OpaqueRef&)
     class OpaqueRef
           ^
```

</details>

**Pull request merged**